### DISCUSSION TITLE
Is this error with MolToSmiles intended behavior?

### DISCUSSION THREAD

**Discussion by claushofmann** [rdkit/rdkit#5371] [Q&A]
Hi everyone,

I am new to Chemoinformatics and was playing around with rdkit for generating different random SMILES strings for the same molecule:

```
from rdkit import Chem
smiles = 'O=C([O-])c1ccccc1O.O=C([O-])c1ccccc1O.[Zn+2]'
mol = Chem.MolFromSmiles(smiles)
Chem.MolToSmiles(mol, doRandom=True, canonical=False)
```

For this specific molecule, I get this error:

```
RuntimeError: Range Error
	idx
	Violation occurred on line 206 in file Code/GraphMol/ROMol.cpp
	Failed Expression: 17 < 1
	RDKIT: 2022.03.3
	BOOST: 1_75
```

Here is a Google Colab demonstrating the issue:

https://colab.research.google.com/drive/1fCZ9rOvwFFMGG_jnp3Scqp4suOoqgQwU#scrollTo=T4jadjXcVAfh

Though I'm not entirely sure, as the error originates from a C++ file, am I right assuming that this is a bug in the library?

**Comment by bp-kelley** [comment]
This is bug like, but the fix on your end is easy.  You have actually have three molecules here and using the doRandom feature gives the canonicalizer conniptions.

To fix this, just split the string on the "."

smiles = 'O=C([O-])c1ccccc1O.O=C([O-])c1ccccc1O.[Zn+2]'

```
>>> smiles = 'O=C([O-])c1ccccc1O.O=C([O-])c1ccccc1O.[Zn+2]'
>>> for smi in smiles.split("."):
...     print(Chem.MolToSmiles(Chem.MolFromSmiles(smi), doRandom=True))
... 
c1cccc(C(=O)[O-])c1O
c1ccc(O)c(c1)C(=O)[O-]
[Zn+2]
```

A small test case for this is:

```
>>> m = Chem.MolFromSmiles("C.C")
>>> Chem.MolToSmiles(m, doRandom=True)
[13:40:09] 

****
Range Error
idx
Violation occurred on line 206 in file /Users/runner/miniforge3/conda-bld/rdkit_1654409106375/work/Code/GraphMol/ROMol.cpp
Failed Expression: 1 < 1
****

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Range Error
	idx
	Violation occurred on line 206 in file Code/GraphMol/ROMol.cpp
	Failed Expression: 1 < 1
	RDKIT: 2022.03.3
	BOOST: 1_7
```

**Reply by bp-kelley** [reply]
I'm not really sure I would classify this as a bug really, as I don't understand what it means to randomize multiple fragments unless it is randomizing the fragments independently.  It's up to @greglandrum to see if we want to fix it or have a better error message at least.

**Reply by claushofmann** [reply]
Ahhh, that makes a lot of sense! Thanks for the explanation and for the fix!

**Reply by claushofmann** [reply]
So, if I understand it correctly the hole thing is called a structure and the individual parts between the `.` are called fragments.

Assuming that any permutation of the fragments in the SMILES string would yield the same structure, I now randomized the string like this:

```
from random import shuffle

randomized_fragments = []
for fragment in smiles.split('.'):
  mol = Chem.MolFromSmiles(fragment)
  randomized_fragments.append(Chem.MolToSmiles(mol, doRandom=True, canonical=False))
shuffle(randomized_fragments)
strcuture = '.'.join(randomized_fragments)
```

**Reply by claushofmann** [reply]
At second thought, I'm not so sure anymore if my assumption that any permutation of the fragments in the SMILES string would yield the same structure is actually correct. 🤔

**Reply by thegodone** [reply]
each SMILES random is equal to the original one MOLECULE (it's a transformer if you speak AI). But do the shuffle will imply that you don't know anymore which is the original one... So it depends of what you need to do with your dataset.

**Comment by bp-kelley** [comment]
I would also say that having multiple fragments makes the complexity of the randomizer considerably higher to sample all variants.

**Comment by greglandrum** [comment]
It's an edge case, but I think it's definitely a bug

**Comment by thegodone** [comment]
Hi guys,

Randomization of a molecule "SMILES" is as Bryan said valid molecule by molecule in theory. 

It should be not returning an error but a warning: cannot apply to mixture aka "." aka lists of SMILES. 

anyway you can implement this little hint, and there are two ways to augment naively such mixtures of SMILES:
1. Same position as initial molecule smiles + doRandom aka Bryan code
2. Adding a first step to code 1 by shuffling list of SMILES (order permutation) and apply doRandom for each of independent SMILES see code 1

BR

Guillaume
